AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/SubscriptionSync/ReconcileAndLocalSave/ReconcileAndLocalSaveOperation.swift

```diff
@@ -137,186 +116,269 @@ class ReconcileAndLocalSaveOperation: AsynchronousOperation {
 
-    /// Execution method for the `applyRemoteModel` disposition. Does not notify directly, but delegates to save or
-    /// delete methods, which eventually notify with:
-    /// - applied
-    /// - errored
-    private func apply(remoteModel: RemoteModel, mutationType: MutationEvent.MutationType) {
-        if log.logLevel == .verbose {
-            log.verbose("\(#function): remoteModel")
-        }
+    func queryLocalMetadata(_ remoteModels: [RemoteModel]) -> Future<([RemoteModel], [LocalMetadata]), DataStoreError> {
+        Future<([RemoteModel], [LocalMetadata]), DataStoreError> { promise in
+            guard !self.isCancelled else {
+                self.log.info("\(#function) - cancelled, aborting")
+                return
+            }
+            guard let storageAdapter = self.storageAdapter else {
+                promise(.failure(DataStoreError.nilStorageAdapter()))
+                return
+            }
 
-        guard !isCancelled else {
-            log.verbose("\(#function) - cancelled, aborting")
-            return
+            guard !remoteModels.isEmpty else {
+                promise(.success(([], [])))
+                return
+            }
+
+            do {
+                let localMetadatas = try storageAdapter.queryMutationSyncMetadata(
+                    forModelIds: remoteModels.map { $0.model.id })
+                promise(.success((remoteModels, localMetadatas)))
+            } catch {
+                promise(.failure(DataStoreError(error: error)))
+                return
+            }
         }
+    }
 
-        guard let storageAdapter = storageAdapter else {
-            Amplify.Logging.log.warn("No storageAdapter, aborting")
-            return
+    func reconcile(_ remoteModels: [RemoteModel],
+                   localMetadatas: [LocalMetadata]) -> [RemoteSyncReconciler.Disposition] {
+        guard let remoteModel = remoteModels.first else {
+            return []
         }
 
-        // TODO: Wrap this in a transaction
-        if mutationType == .delete {
-            saveDeleteMutation(storageAdapter: storageAdapter, remoteModel: remoteModel)
-        } else {
-            saveCreateOrUpdateMutation(storageAdapter: storageAdapter,
-                                       remoteModel: remoteModel,
-                                       mutationType: mutationType)
+        let dispositions = RemoteSyncReconciler.reconcile(remoteModels,
+                                                          localMetadatas: localMetadatas)
+        for _ in 0 ..< (remoteModels.count - dispositions.count) {
+            notifyDropped(modelName: remoteModel.model.modelName)
         }
+
+        return dispositions
     }
 
-    private func saveDeleteMutation(storageAdapter: StorageEngineAdapter,
-                                    remoteModel: RemoteModel) {
-        log.verbose(#function)
+    func applyRemoteModels(dispositions: [RemoteSyncReconciler.Disposition]) -> Future<Void, DataStoreError> {
```

Comment:
As discussed, there's a log going on in this method, add a TODO to refactor in a future PR. Although there is a bit of code dupe in the body of the `switch`, I see its purpose and it's probably not worth the effort to abstract it. It would be nice to move each of the publisher constructions to its own utility method though for readability of the `switch`, and a single method that you can invoke in the `map`: